### DISCUSSION TITLE
How to enable beginner users to use external libraries

### DISCUSSION THREAD

**Discussion by tatsuya-takakuwa** [biolab/orange3#6628] [General]
I want to allow beginner users to use a Python script that incorporates an external library. Is it possible to provide this to the users without requiring them to write code?

**Comment by chrislee35** [comment]
Mr. Takukuwa, there are several ways to use Python libraries with an external library, but they all require some code. Please tell me what you are trying to do.  Your ideas will inspire a guidebook on how to do several things with minimal code.
たくわさん、Python ライブラリを外部ライブラリで使用する方法はいくつかありますが、いずれもコードが必要です。 何をしようとしているのか教えてください。 あなたのアイデアは、最小限のコードでいくつかのことを行う方法に関するガイドブックのインスピレーションになります。
Recently, I found a way to include new functions in the Formula widget.  I give an example with datetime variable type.
最近、数式ウィジェットに新しい関数を含める方法を見つけました。 (datetime)変数型の例を示します。
```
from Orange.widgets.data.owfeatureconstructor import __GLOBALS, FeatureEditor

import datetime
import zoneinfo

def fts(args):
    if type(args) == list:
        return datetime.datetime.fromtimestamp(*args)
    return datetime.datetime.fromtimestamp(args)

callbacks = [
    ["timezone", lambda args: zoneinfo.ZoneInfo(args) ],
    ["year", lambda args: fts(args).year ],
    ["month", lambda args: fts(args).month ],
    ["day_of_month", lambda args: fts(args).day ],
    ["day_of_week", lambda args: fts(args).weekday() ],
    ["day_of_year", lambda args: fts(args).timetuple().tm_yday ],
    ["week_of_year", lambda args: fts(args).timetuple().tm_yday/7.0 ],
    ["quarter", lambda args: fts(args).month // 4],
]

for name, cb in callbacks:
    __GLOBALS[name] = cb
    FeatureEditor.FUNCTIONS[name] = cb
```
I hope this is helpful to you. これがお役に立てば幸いです。

**Comment by markotoplak** [comment]
@chrislee35, this is quite a trick. Be aware that this touches the internals deeply, so it can stop working at upgrades. On the other hand, it will likely keep working for quite some time...

**Comment by chrislee35** [comment]
For this one off example, I'm okay with that.  If I extend this to include other mapper/calculation functions (e.g., IP addressses), I might patch owfeatureconstructor to have a singleton FunctionsManager.  The FunctionsManager could have functions added through add-ons, settings, or other avenues, to then make that available to all instances of the Formula widget.  But, I haven't matured this idea enough to officially propose it.

**Comment by wvdvegte** [comment]
I'm wondering what this request and discussion is about. How can it make sense to use the Python Script widget without writing code? And if you use Python Script, it is already possible to use an external library, for instance you can start the script with "import pandas" ...

**Comment by janezd** [comment]
Thanks @wvdvegte, I'm rather confused as well about what's happening here. :)

Anyway, I'm converting it to a discussion and let it live.

By the way, @chrislee35, I like your idea about add-ons registering new functions for feature constructor. At least in principle -- if it doesn't cause problems.
